add optional support for "file-magic"

FelixSchwarz opened the pull request:
There are multiple Python packages for the "libmagic" library. Unfortunately, some of them use the same Python package name ("magic") and can not be installed in parallel.

This code adds support for the "file-magic" pypi package as a fallback when "python-magic" is not installed. Unfortunately `pyproject.toml` does not provide a way to express a XOR dependency so regular users will always use "python-magic". However this change enables Linux distributions like Fedora to simply patch `pyproject.toml`.

In particular, Fedora's file manager "dnf" has an indirect dependency on `python3-file-magic` so `python-magic` can not be installed. Not sure if you want to take that patch but it would certainly simplify packaging for Fedora. However I can also just carry this change as a downstream patch.

yaroslaff commented:
If I understand correctly, when showcert is installed using pip or pipx, it will work as always (using python-magic), nothing changes in this case. But this conditional code will be used only if installed from Fedora-patched versions with file-magic?

FelixSchwarz commented:
exactly.

In Fedora I added a [patch to remove `python-magic` from `pyproject.toml`](https://src.fedoraproject.org/rpms/showcert/blob/4359800bf9638d9917aee8ad7c23f10cb267da31/f/showcert-remove-dependency-on-python-magic.patch) and instead [require `file-magic` via RPM requirements](https://src.fedoraproject.org/rpms/showcert/blob/4359800bf9638d9917aee8ad7c23f10cb267da31/f/showcert.spec#_18).

This means a Fedora user will install either `python3-magic` or `python3-file-magic` (in practice always the latter due to the dnf dependency). As the patched showcert code can handle both implementations, everything will be fine.

yaroslaff commented:
I like this idea but "NameError: name 'rawcert' is not defined" and pytest fails almost all tests:

```
def _detect_mimetype(file_contents: bytes) -> str:
    if hasattr(magic, "from_buffer"):
        # https://pypi.org/project/python-magic/
        return magic.from_buffer(rawcert, mime=True)
    # https://pypi.org/project/file-magic/
    _file_magic = magic.detect_from_content(file_contents)
    return _file_magic.mime_type
```

FelixSchwarz committed:
add optional support for "file-magic"

There are multiple Python packages for the "libmagic" library. Unfortunately,
some of them use the same Python package name ("magic") and can not be installed
in parallel.

This code adds support for the "file-magic" pypi package as a fallback when
"python-magic" is not installed. Unfortunately `pyproject.toml` does not provide
a way to express a XOR dependency so regular users will always use "python-magic".
However this change enables Linux distributions like Fedora to simply patch
`pyproject.toml`.

FelixSchwarz committed:
read cert from `sys.stdin.buffer` as bytes

FelixSchwarz commented:
🤦
I updated the patch + add a read from `sys.stdin.buffer` so that `_detect_mimetype()` always gets bytes. I can drop the second patch if you feel it is over the top.

yaroslaff commented:
merged, thank you!

FelixSchwarz commented:
Thank you for catching that embarassing mistake + merging it so quickly.
